File: frontend/src/utils/prescriptionSchedule.js
/** Shared prescription schedule / food-timing options (consultation + inpatient). */

export const FREQUENCY_OPTIONS = [
  { value: '1-0-0', label: 'Morning only' },
  { value: '0-1-0', label: 'Afternoon only' },
  { value: '0-0-1', label: 'Night only' },
  { value: '1-0-1', label: 'Morning & Night' },
  { value: '1-1-0', label: 'Morning & Afternoon' },
  { value: '1-1-1', label: 'Three times a day' },
  { value: '0-1-1', label: 'Afternoon & Night' },
];

export const FOOD_TIMING_OPTIONS = [
  { value: 'before_food', label: 'Before food' },
  { value: 'after_food', label: 'After food' },
  { value: 'with_food', label: 'With food' },
  { value: 'on_empty_stomach', label: 'Empty stomach' },
  { value: 'anytime', label: 'Anytime' },
];

const FOOD_TIMING_LABELS = Object.fromEntries(
  FOOD_TIMING_OPTIONS.map((o) => [o.value, o.label.toLowerCase()]),
);

/** Human-readable dosage line matching consultation / OPD prescription PDFs. */
export function buildDosageInstruction(dosage, frequencySchedule = '1-0-0', foodTiming = 'after_food') {
  const schedule = frequencySchedule || '1-0-0';
  const [morning, afternoon, night] = schedule.split('-');
  const timings = [];
  if (morning === '1') timings.push('morning');
  if (afternoon === '1') timings.push('afternoon');
  if (night === '1') timings.push('night');
  const frequencyText = timings.length > 0 ? timings.join(', ') : 'once daily';
  const foodLabel = FOOD_TIMING_LABELS[foodTiming] || 'after food';
  return `${dosage || '1 dose'} - ${frequencyText} ${foodLabel}`;
}

export const BLANK_INPATIENT_RX_ITEM = {
  medicine_id: '',
  medicine_name: '',
  dosage: '',
  frequency_schedule: '1-0-0',
  food_timing: 'after_food',
  duration: '',
  quantity_prescribed: 1,
  instructions: '',
};

/** Take-home meds on discharge — same schedule fields as consultation Rx. */
export const BLANK_TAKE_HOME_MED = {
  medicine_id: '',
  medicine_name: '',
  dosage: '',
  frequency_schedule: '1-0-0',
  food_timing: 'after_food',
  duration: '',
  quantity: '',
  instructions: '',
};

export function frequencyScheduleLabel(schedule = '1-0-0') {
  return FREQUENCY_OPTIONS.find((o) => o.value === schedule)?.label || schedule;
}

export function foodTimingLabel(timing = 'after_food') {
  return FOOD_TIMING_OPTIONS.find((o) => o.value === timing)?.label || timing;
}

/** Map a take-home med form row to the discharge API payload. */
export function serializeTakeHomeMed(m) {
  const schedule = m.frequency_schedule || '1-0-0';
  const food = m.food_timing || 'after_food';
  const extra = m.instructions?.trim() || '';
  const foodLabel = foodTimingLabel(food);
  return {
    medicine_id: m.medicine_id ? parseInt(m.medicine_id, 10) : null,
    medicine_name: m.medicine_name.trim(),
    dosage: m.dosage?.trim() || null,
    frequency: frequencyScheduleLabel(schedule),
    duration: m.duration?.trim() || null,
    quantity: m.quantity ? parseInt(m.quantity, 10) : null,
    instructions: extra ? `${foodLabel}. ${extra}` : foodLabel,
    frequency_schedule: schedule,
    food_timing: food,
  };
}

File: backend/app/utils/prescription_schedule.py
"""Map consultation-style frequency_schedule codes to MAR fields."""

from __future__ import annotations

import re
from typing import Optional, Tuple, List

FOOD_TIMING_LABELS = {
    "before_food": "before food",
    "after_food": "after food",
    "with_food": "with food",
    "on_empty_stomach": "on empty stomach",
    "anytime": "anytime",
}

# frequency_schedule → (frequency code for MAR, explicit HH:MM times)
_SCHEDULE_TO_MAR: dict[str, Tuple[Optional[str], List[str]]] = {
    "1-0-0": ("OD", ["09:00"]),
    "0-1-0": (None, ["14:00"]),
    "0-0-1": ("HS", ["22:00"]),
    "1-0-1": ("BD", ["08:00", "20:00"]),
    "1-1-0": (None, ["08:00", "14:00"]),
    "1-1-1": ("TDS", ["08:00", "14:00", "20:00"]),
    "0-1-1": (None, ["14:00", "20:00"]),
}


def schedule_to_mar(frequency_schedule: Optional[str]) -> Tuple[Optional[str], List[str]]:
    key = (frequency_schedule or "1-0-0").strip()
    return _SCHEDULE_TO_MAR.get(key, ("OD", ["09:00"]))


def build_dosage_instruction(
    dosage: str,
    frequency_schedule: Optional[str] = "1-0-0",
    food_timing: Optional[str] = "after_food",
) -> str:
    schedule = (frequency_schedule or "1-0-0").strip()
    parts = schedule.split("-")
    morning = parts[0] if len(parts) > 0 else "0"
    afternoon = parts[1] if len(parts) > 1 else "0"
    night = parts[2] if len(parts) > 2 else "0"
    timings = []
    if morning == "1":
        timings.append("morning")
    if afternoon == "1":
        timings.append("afternoon")
    if night == "1":
        timings.append("night")
    frequency_text = ", ".join(timings) if timings else "once daily"
    food_label = FOOD_TIMING_LABELS.get(food_timing or "after_food", "after food")
    dose = (dosage or "1 dose").strip()
    return f"{dose} - {frequency_text} {food_label}"


def parse_duration_days(duration: Optional[str]) -> Optional[int]:
    if not duration:
        return None
    m = re.search(r"(\d+)", duration)
    if not m:
        return None
    try:
        return int(m.group(1))
    except ValueError:
        return None
